### ISSUE TITLE
Load 20 images per time will lost some images

### ISSUE THREAD

**Issue by LuongTruong** [react-native-community/react-native-cameraroll#47]
Hi everyone, I am using CameraRoll to implement the image Gallery and there is a bug

Bug details
========

I load 20 images first, after user scroll down to bottom, another 20 images load. This thing is: 
- I have total 201 images.
- If I load 1000000 images, I can get all 201 images.
- If I load 20 or 10 images per time, I only receive 190+ images.

To Reproduce
==========

I create a repo so that you guys can have a better investigation.

[Camera Roll Bug](https://github.com/truongluong1314520/RNCameraRollBug)

You can download or clone the repo, "npm install", and "react-native run-android"

Expected Behavior
==============

Lazy load or load all images, I can still get all of the images from the phone.

Environment
=========

React Native Environment Info:
    System:
      OS: macOS 10.14.1
      CPU: (12) x64 Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
      Memory: 2.64 GB / 16.00 GB
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 10.15.0 - /usr/local/bin/node
      Yarn: 1.13.0 - /usr/local/bin/yarn
      npm: 6.9.0 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 12.1, macOS 10.14, tvOS 12.1, watchOS 5.1
      Android SDK:
        API Levels: 16, 17, 18, 19, 21, 22, 23, 24, 25, 26, 27, 28
        Build Tools: 23.0.1, 25.0.2, 26.0.1, 26.0.2, 26.0.3, 27.0.3, 28.0.2, 28.0.3
        System Images: android-19 | Google APIs Intel x86 Atom, android-21 | Google APIs Intel x86 Atom, android-22 | Google APIs Intel x86 Atom, android-23 | Google APIs Intel x86 Atom, android-24 | Google Play Intel x86 Atom, android-25 | Google Play Intel x86 Atom, android-26 | Google Play Intel x86 Atom, android-27 | Google Play Intel x86 Atom, android-28 | Google APIs Intel x86 Atom, android-28 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: 3.1 AI-173.4907809
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.6.0-alpha.8af6728 => 16.6.0-alpha.8af6728 
      react-native: 0.57.4 => 0.57.4 
    npmGlobalPackages:
      react-native-cli: 2.0.1

**Comment by bartolkaruza** [comment]
Thanks for the clear report, we'll have a look at this!

**Comment by truongluong1314520** [comment]
Hi @bartolkaruza, is there any news for this issue?

**Comment by jonathangreco** [comment]
Hi @bartolkaruza Some time has passed and still no answer on this, did you manage to reproduce this issue ?

**Comment by jonathangreco** [comment]
@bartolkaruza Please an answer on this one thanks in advance

**Comment by bartolkaruza** [comment]
@javache any chance you could help out on this issue? There is a link to a repro in the description.

**Comment by jonathangreco** [comment]
I would gladly appreciate the confirmation of my issue @LuongTruong , what happens if you have less than 10 images in your gallery ? and ask for 20 ? 

In my project no images are shown.

**Comment by nihp** [comment]
I am also using cameraroll to get all photos. I need to load 40 and then 40 images. 

It not fetching all images. First 40 images are only visible all the other images are not shown in UI. 

Sometimes it fetching all images but not showing the images loading it loading all the images again and again. So it shows as flickering of the grid

**Comment by thekevinbrown** [comment]
I think we're experiencing this with a user on iOS. I'll have a look and see if we can fix it.

**Comment by thekevinbrown** [comment]
The problem on Android is we're using Date Taken as a pagination mechanism, but there's no guarantee that photos will have a different timestamp:

```
2020-01-20 15:37:03.216 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000542458.jpg, Date Taken: 1566866253000
2020-01-20 15:37:03.217 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000553233.jpg, Date Taken: 1566866253000
2020-01-20 15:37:03.217 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000028462.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.217 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000040912.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.217 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000065562.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.218 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000083959.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.218 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000209400.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.218 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000258129.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.218 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000302227.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.219 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000381890.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.219 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000394523.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.219 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000407676.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.220 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000413603.jpg, Date Taken: 1566866252000
2020-01-20 15:37:03.220 27527-27703/com.camerarolllazyload I/CameraRoll: Image ID: file:///storage/emulated/0/Pictures/train2014/COCO_train2014_000000439940.jpg, Date Taken: 1566866252000
```

So when we end up with 20 photos on the same timestamp, the page info quite happily tells us to only fetch photos older than, say `1566866252000`, but of course that skips more photos than it is meant to. Working on a patch for the Java/Android version of this issue now, then will look into iOS.

**Comment by thekevinbrown** [comment]
Ok, on iOS the implementation looks deterministic at least, but it also looks like it'd perform pretty terribly because it does a scan through the photos until it finds one that matches a URI, then it continues for the page size and exits.

@bartolkaruza Would you guys accept a PR to improve the performance here that uses [objectsAtIndexes:](https://developer.apple.com/documentation/photokit/phfetchresult/1621008-objects) to pull a single page out at a time? We'd then switch over to the new Android approach that effectively uses offset and limit, but this way you don't have to scan through the entire camera roll for pages towards the end of the result set.

**Comment by bartolkaruza** [comment]
@thekevinbrown yes, that sounds like a much needed improvement. I suspect it will fix (or improve) #83
